### PR TITLE
[public-api-server] Group http and grpc servers into a baseserver package

### PR DESCRIPTION
**Pull request by easyCZ**
## Description

Subsequent PRs will add (as needed):
* Cert support
* Metrics collector and endpoint
* Health/Ready endpoints
* Additional configuration as required

This change is primarily for WebApp, but Github tags Workspace also due to a new library in common-go. Unfortunately, this change is bigger than my regular changes but it needed to remain together to show a working example.

## Related Issue(s)

Part of https://github.com/gitpod-io/gitpod/issues/9229

## How to test

```
cd components/common-go/baseserver && go test ./...
```

```
➜ cd components/public-api-server 
➜ go run main.go 
INFO[0000] Serving gRPC on [::]:9001                     protocol=grpc
INFO[0000] Serving http on [::]:9000                     protocol=http
^CINFO[0004] Received system signal interrupt, closing server. 
INFO[0004] Received graceful shutdown request.          
INFO[0004] GRPC server terminated.                      
INFO[0004] HTTP server terminated.     
```

## Release Notes

```release-note
NONE
```

## Documentation

NONE

/hold

### CODE REVIEW

**Comment by roboquat** [comment]
@easyCZ: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by easyCZ** [comment]
Werft build keeps flaking, but this can be reviewed and I'll re-run the build at a later point. The preview build does work.

File: components/common-go/baseserver/server.go
Comment (andrew-farries) [inline components/common-go/baseserver/server.go:151]: nit: typo here (operaiton)

File: components/common-go/baseserver/server.go
Comment (andrew-farries) [inline components/common-go/baseserver/server.go:101]: I don't think you need to specify a channel buffer size here.

File: components/common-go/baseserver/server.go
Comment (andrew-farries) [inline components/common-go/baseserver/server.go:86]: Why does this server (and the http one below) need to be started in a goroutine? Could we not simply error out if we fail to start either of them without having them send their errors on a channel only to collect them on line 105?

**Review by andrew-farries — CHANGES_REQUESTED** [review]

File: components/common-go/baseserver/server.go
Comment (easyCZ) [reply components/common-go/baseserver/server.go:86]: Needs to be a go-routine because grpc.Serve and `http.ListenAndServe` are blocking. This follows the standard golang practice where if you want things to be in the background, the caller can do that themselves.

We need to put them in go-routines because we need to start both of them, and then synchronously await termination of either (or both) of the servers before exiting to maintain the semantics of listen and serve.

File: components/common-go/baseserver/options_test.go
Comment (easyCZ) [reply components/common-go/baseserver/options_test.go:14]: Sure, happy to add them.

File: components/common-go/baseserver/server.go
Comment (easyCZ) [reply components/common-go/baseserver/server.go:101]: I was going by this example: https://gobyexample.com/signals

I think it doesn't hurt to specify it as that we're ensuring that repeated signals (unsure if we'd receive them) don't get buffered and block instead. That's desirable as we're looking to limit memory usage and complete our termination procedure. In practice, in k8s, we'll receive SIGINT, and if we don't shut down in time k8s will just terminate the pod entirely

File: components/common-go/baseserver/options_test.go
Comment (easyCZ) [reply components/common-go/baseserver/options_test.go:14]: Added tests and pushed.

**Review by andrew-farries — DISMISSED** [review]

**Comment by princerachit** [comment]
@easyCZ would it make sense if workspace team approves this change only after webapp has approved? As you mentioned the changes are related to webapp and we are only getting tagged for a new library in common.go

**Comment by princerachit** [comment]
@easyCZ would it make sense if workspace team approves this change only after webapp has approved? As you mentioned the changes are related to webapp and we are only getting tagged for a new library in common.go

**Comment by easyCZ** [comment]
@princerachit Thanks for taking a look. @andrew-farries previously approved but I had to rebase due to a downstream change. If you're happy with the changes, I'll put on a hold. In this change, both approvals (WebApp, Workspace) so even with your approval, it won't get auto-merged.

/hold

**Review by andrew-farries — APPROVED** [review]

File: components/common-go/baseserver/options.go
Comment (princerachit) [inline components/common-go/baseserver/options.go:60]: nit: Should it be ?
```suggestion
		if port <= 0 {
```

File: components/common-go/baseserver/options.go
Comment (princerachit) [inline components/common-go/baseserver/options.go:72]: maybe?
```suggestion
			return fmt.Errorf("no logger specified")
```

**Review by princerachit — APPROVED** [review]

File: components/common-go/baseserver/options.go
Comment (easyCZ) [reply components/common-go/baseserver/options.go:60]: This is actually a bug in the HTTP Port option. When you specify port `0`, `net.Listen` actually allocates an open port. I'll fix this in a follow-up PR to avoid tangling you in the review again. Thanks for flagging!

File: components/common-go/baseserver/options.go
Comment (easyCZ) [reply components/common-go/baseserver/options.go:72]: Intent was to be more developer friendly and actually tell them it's nil when we know it is.

**Pull request merged**
